src/index.js

```diff
@@ -126,30 +127,58 @@ class Node extends EventEmitter {
     this.peerInfo.multiaddrs.replace(maOld, maNew)
 
     const multiaddrs = this.peerInfo.multiaddrs.toArray()
-    transports.forEach((transport) => {
-      if (transport.filter(multiaddrs).length > 0) {
-        this.switch.transport.add(
-          transport.tag || transport.constructor.name, transport)
-      } else if (WebSockets.isWebSockets(transport)) {
-        // TODO find a cleaner way to signal that a transport is always
-        // used for dialing, even if no listener
-        ws = transport
+
+    this._modules.transport.forEach((Transport) => {
+      let t
+
+      if (typeof Transport === 'function') {
+        t = new Transport()
+      } else {
+        t = Transport
       }
+
+      if (t.filter(multiaddrs).length > 0) {
+        this._switch.transport.add(t.tag || t.constructor.name, t)
+      } else if (WebSockets.isWebSockets(t)) {
+        // TODO find a cleaner way to signal that a transport is always used
+        // for dialing, even if no listener
+        ws = t
+      }
+      this._transport.push(t)
     })
 
     series([
-      (cb) => this.switch.start(cb),
+      (cb) => this._switch.start(cb),
       (cb) => {
         if (ws) {
           // always add dialing on websockets
-          this.switch.transport.add(ws.tag || ws.constructor.name, ws)
+          this._switch.transport.add(ws.tag || ws.constructor.name, ws)
         }
 
         // all transports need to be setup before discover starts
-        if (this.modules.discovery) {
-          return each(this.modules.discovery, (d, cb) => d.start(cb), cb)
+        if (this._modules.peerDiscovery && this._config.peerDiscovery) {
+          each(this._modules.peerDiscovery, (D, cb) => {
+            // If enabled then start it
+            if (this._config.peerDiscovery[D.tag].enabled) {
```

Comment:
We need to also verify the existence of `this._config.peerDiscovery[D.tag]`. This will break when that is inevitably missing from the config. 

This brings up a question around defaults if the config is not present. Does the config **have** to exist and be **enabled** in order for the discovery service to be added?

I think the logic should be:
* If `_modules.peerDiscovery` contains an instance, use that and ignore info in `_config.peerDiscovery`. If I pass an instance I'd expect that to be enabled and used as is.
* If `_modules.peerDiscovery` contains a function, its configuration exists in `_config.peerDiscovery` and isn't explicitly `enabled: false`, it is added. Otherwise, it will be ignored.

I think this helps make configuration easier. If I supply something to modules I want it enabled unless I explicitly disable it.

Reply:
Let's rather use a strategy such as https://github.com/ipfs/js-ipfs/blob/master/src/core/config.js#L5-L41 to ensure that the config exists, letting it represent the truth about _the configuration_. There are so many options that it is better to be explicit rather than creating too many ternaries.